kore/src/Kore/Syntax/Json/Internal.hs

```diff
@@ -0,0 +1,687 @@
+        , unwrapUnaryRecords = True
+        , tagSingleConstructors = True
+        , rejectUnknownFields = True
+        }
+  where
+    constructorTagModifier = \case
+        "KJDv" -> "dv"
+        'K' : 'J' : rest -> rest
+        other -> other
+
+----------------------------------------
+-- Identifiers and lexical checks
+
+newtype Id = Id Text
+    deriving stock (Eq, Show, Generic)
+    deriving newtype (ToJSON, FromJSON)
+
+{- | Performs a (shallow, top-level, no recursion) lexical check of
+ identifiers contained in the given node. For details see the check
+ functions below.
+-}
+lexicalCheck :: KorePattern -> Json.Parser KorePattern
+lexicalCheck p =
+    case p of
+        KJEVar{name = Id n} ->
+            reportErrors n "element variable" checkIdChars
+        KJSVar{name = Id n} ->
+            reportErrors n "set variable" checkSVarName
+        KJApp{name = Id n} ->
+            reportErrors n "app symbol" checkSymbolName
+        -- KJString txt ->
+        --     reportErrors txt "string literal" checkStringChars
+        -- for the time being, use standard Text (supports std
+        -- Unicode) and do not check.
+        KJForall{var = Id name} ->
+            reportErrors name "quantifier variable" checkIdChars
+        KJExists{var = Id name} ->
+            reportErrors name "quantifier variable" checkIdChars
+        KJMu{var = Id name} ->
+            reportErrors name "fixpoint expression variable" checkSVarName
+        KJNu{var = Id name} ->
+            reportErrors name "fixpoint expression variable" checkSVarName
+        KJDv{value = txt} ->
+            reportErrors txt "domain value string" checkStringChars
+        KJMultiApp{symbol = Id n} ->
+            reportErrors n "multi-app symbol" checkSymbolName
+        _ -> pure p
+  where
+    reportErrors :: Text -> String -> (Text -> [String]) -> Json.Parser KorePattern
+    reportErrors text kind check
+        | null errors = pure p
+        | otherwise =
+            fail . unwords $
+                [ "Lexical"
+                , if length errors == 1 then "error" else "errors"
+                , "in"
+                , kind
+                , ":"
+                , T.unpack text
+                ]
+                    <> map ("* " <>) errors
+      where
+        errors = check text
+
+{- | Basic identifiers start with letters and may contain letters,
+ digits, _ or '. Set variables start with '@' followed by a basic
+ identifier. Symbol variables _may_ start by \, followed by a basic
+ identifier.
+-}
+checkIdChars :: Text -> [String] -- list of lexical errors
+checkIdChars name
+    | T.null name = ["Empty"]
+checkIdChars name =
+    ["Illegal initial character " <> show first | not $ isAlpha first]
+    ++ [ "Contains illegal characters: " <> show (nub $ T.unpack illegalChars)
+       | not $ T.null illegalChars
+       ]
+  where
+    first = T.head name
+    illegalChars = T.filter (not . isIdChar) $ T.tail name
```

Comment:
BTW the `null` case is separate here because the `T.head` and `T.tail` in the `where` clause will crash the program otherwise. I don't like strict Haskell.

Reply:
ah yeah, I've fixed a bug due to this about two months ago 🥲

Reply:
Probably a silly question: would making the bindings lazy (`~first` and `~illegalChars`) work here?